docker/image.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi";
+import { buildAndPushImage, DockerBuild, Registry } from "./docker";
+
+/**
+ * Arguments for constructing an Image resource.
+ */
+export interface ImageArgs {
+    /**
+     * The qualified image name that will be pushed to the remote registry.  Must be a supported image name for the
+     * target registry user.
+     */
+    imageName: pulumi.Input<string>;
+    /**
+     * The Docker build context, as a folder path or a detailed DockerBuild object.
+     */
+    build: pulumi.Input<string | DockerBuild>;
+    /**
+     * The docker image name to build locally before tagging with imageName.
+     */
+    localImageName?: pulumi.Input<string>;
+    /**
+     * Docker registry to push to.
+     */
+    registry: pulumi.Input<ImageRegistry>;
+}
+
+export interface ImageRegistry {
+    /**
+     * Docker registry to push to.
+     */
+    name: pulumi.Input<string>;
+    /**
+     * Username for login to the target Docker registry.
+     */
+    username: pulumi.Input<string>;
+    /**
+     * Password for login to the target Docker registry.
+     */
+    password: pulumi.Input<string>;
```

Comment:
Just a note, perfect example where we want a Secret<T>.

Reply:
:+1:  

Interestingly, because we are making this a component now, and are passing these values as inputs to the component, the value of `password` will get stored in the checkpoint file now - even though it didn't in the non-component approach previously.  Notably, in the AWS and GCR cases, these passwords are retrieved as short-lived credentials.